### PR TITLE
[RFC] Mcuboot XIP Support for Zephyr

### PR DESCRIPTION
**Pull request by de-nordic**
Provided script will automatically build alternative slot application.

Invocation:
python3 zephyr/scripts/direct_xip.py <original_build>

TODO:
- [ ] add destination path selection instead of using harcoded one
- [x] error/exception handling

### CODE REVIEW

**Comment by de-nordic** [comment]
Mcuboot says it works, smp_svr sample executed from slot 2
```
uart:~$ *** Booting Zephyr OS build zephyr-v2.4.0-1312-g2f4b66967d31  ***
[00:00:00.256,378] <inf> mcuboot: Starting bootloader
[00:00:00.262,329] <inf> mcuboot: Primary slot: Image not found
[00:00:00.268,951] <inf> mcuboot: Secondary slot: version=0.0.0+0
[00:00:00.416,137] <inf> mcuboot: Booting image from the secondary slot
[00:00:00.423,492] <inf> mcuboot: Bootloader chainload address offset: 0x73000
[00:00:00.431,488] <inf> mcuboot: Jumping to the first image slot
*** Booting Zephyr OS build zephyr-v2.4.0-1312-g2f4b66967d31  ***
[00:00:00.005,859] <inf> smp_sample: build time: Nov 12 2020 15:10:12
```

Dump from mcumgr:
```
$$ mcumgr_serial image list
Images:
 image=0 slot=1
    version: 0.0.0
    bootable: true
    flags:
    hash: b29b36984a7e4d2a83d4cfe34d53ff96d16bb75b85cdbf24f02ff8d73ec32807
Split status: N/A (0)
```

File: Kconfig.zephyr [hunk 1]
```diff
@@ -80,20 +80,46 @@ config HAS_FLASH_LOAD_OFFSET
...
-	bool "Link application into /chosen/zephyr,code-partition from devicetree"
+	bool "Link application into one of device tree defined partitions"
 	help
-	  When enabled, the application will be linked into the flash partition
-	  selected by the zephyr,code-partition property in /chosen in devicetree.
-	  When this is disabled, the flash load offset and size can be set manually
-	  below.
+	  When enabled, the choice of FLASH_LOAD_OFFSET/FLASH_LOAD_SIZE will
+	  be restricted to fixed partitions defined in device tree.
+	  By default the chisen zephyr,code-partition will be used.
```
Comment (utzig) [inline Kconfig.zephyr:87]: typo: `chisen` -> `chosen`

File: Kconfig.zephyr [hunk 2]
```diff
@@ -80,20 +80,46 @@ config HAS_FLASH_LOAD_OFFSET
...
+DT_IMAGE_0_PARTITION := slot0_partition
+DT_IMAGE_1_PARTITION := slot1_partition
+
+if USE_DT_CODE_PARTITION
+
+choice
+	prompt "Application partition"
+	default APP_IN_Z_CODE_PARTITION
+	help
+	  Menu allows to select parition the application will be build for.
```
Comment (utzig) [inline Kconfig.zephyr:102]: typos: `parition` -> `partition`, `build` -> `built`

File: Kconfig.zephyr [hunk 3]
```diff
@@ -80,20 +80,46 @@ config HAS_FLASH_LOAD_OFFSET
...
+
+config APP_IN_Z_CODE_PARTITION
+	bool "zephyr,code-partition \($(dt_chosen_label,$(DT_CHOSEN_Z_CODE_PARTITION))\)"
+
+config APP_IN_IMAGE_0_PARTITION
+	bool "slot0_parition"
```
Comment (utzig) [inline Kconfig.zephyr:108]: typo: `slot0_parition` -> `slot0_partition`

File: Kconfig.zephyr [hunk 4]
```diff
@@ -450,6 +478,14 @@ config MCUBOOT_GENERATE_CONFIRMED_IMAGE
 	  The existence of bin and hex files depends on CONFIG_BUILD_OUTPUT_BIN
 	  and CONFIG_BUILD_OUTPUT_HEX.
 
+config MCUBOOT_TARGET_IMAGE
+	int"Selects image slot for which application will be build"
```
Comment (utzig) [inline Kconfig.zephyr:482]: add a space after the keyword, typo: `build` -> `built`

**Review by utzig — CHANGES_REQUESTED** [review]

File: Kconfig.zephyr (hunk 4 above)
Comment (de-nordic) [reply Kconfig.zephyr:482]: Option will be remove; this is leftover from experiments.

**Comment by Laczen** [comment]
@de-nordic, I don't think it is required to do this change, correct usage of chosen partition allows to generate the correct image. The chosen partition info also has the required data to be passed to imgtool and to dfu-util/nrfjprog/...
To allow support for a bootloader (not only mcuboot) there are only two required pieces of info for a application image: where it will be placed in flash (partition offset) and where the image starts (unencrypted, uncompressed, ...): this is the partition offset + a image header size.

Easier support for uploading using dfu-util could be achieved by adding the location where the binary should be placed in flash to the header info, in that case the dfu "server" could  read in the header and decide where to place it or to reject it. This is much more flexible that selecting alt-0 or alt-1 (BTW, selecting alt-0 to "upload" from flash should be disallowed). When using a "hex"-loader this is maybe less required, but still usefull for the bootloader to distinguish between execute or move.

**Comment by de-nordic** [comment]
> 
> Easier support for "downloading" using dfu-util could be achieved by adding the location where the binary should be placed in flash to the header info in that case the dfu "server" could read in the header and decide where to place it or to reject it.

Yes, but you are talking abut hi-end solution, I am still trying to build things for two slots. You may add that comment to the original issue reported here: https://github.com/zephyrproject-rtos/zephyr/issues/27673.

But yes, it would be nice to add header that would identify the slot of binary and allow dfu client to reject it citing the problem, or even pack two binaries together (one for each slot) and allow dft to negotiate which one will go in.

I will try with  `zephyr.dfu-code-partition` that is supposed to be set to the slot the application will be uploaded to.
The combination of `zephyr,code-partition` and `zephyr, dfu-code-partition`, or lack of the former one, would decide whether application is built for dual slot bootable device, XIP dual boot device or single slot device.

**Comment by Laczen** [comment]
@de-nordic, maybe I am missing something, I was thinking you are trying to create a image to be executed from slot1. This could be done by setting the chosen partition to slot1 (or image1) nothing more. But then again, maybe you are trying to achieve something else.

**Comment by nandojve** [comment]
Hi folks, I noted this improvement and have couple questions:

1) How MCUboot will handle 3 or 4 images, or even more?

I mean, there are work on Zephyr to enable NS and SE images (non secure and secure images). These are two independent partitions + at least another one as temporary storage (scenario 1). So, how MCUboot + Zephyr will handle this ? I imagine XIP is essential for lots of platforms and there is a special case, i.MX-RT that only have external memory. Having two images on a single bin (scenario 2) is an alternative but can be impractical with LWPAN. My firsts tests with BLE IPSP (PC <-> nRF) needs around 30min to download and update 1 image with minimal resources to run the example without mesh and only 2 nodes.

2) Since there will be needed work to change the current dual image to a new thing, would be interesting have a infrastructure that allows multi-image support, or something that address this current/future scenario?

Sorry if this is not the right place to add both questions but it seemed opportune to raise them.

CC: @otavio

**Comment by utzig** [comment]
> 1. How MCUboot will handle 3 or 4 images, or even more?

The `mcumgr` CLI app already has a flag to choose the image pair where to upload, `image upload -n <image-number>`, which can be used to choose between S/NS, but this currently uploads in a fixed slot. I don't remember if the mcumgr library supports this, but MCUboot surely does. With the XIP support it will probably also get an extra flag for the slot number. Then a user will have control of what image pair and slot a particular image should be uploaded to.

File: subsys/usb/class/usb_dfu.c [hunk 5]
```diff
@@ -344,7 +353,7 @@ static void dfu_reset_counters(void)
 {
 	dfu_data.bytes_sent = 0U;
 	dfu_data.block_nr = 0U;
-	if (flash_img_init(&dfu_data.ctx)) {
+	if (flash_img_init_id(&dfu_data.ctx, UPDATEABLE_FLASH_IMG_ID)) {
```
Comment (nvlsianpu) [inline subsys/usb/class/usb_dfu.c:356]: why this is not applied in `dfu_data declaration`?

**Review by nvlsianpu — COMMENTED** [review]
Did you consider to keep working usb dfu class when it is part of mcuboot in single application-slot-mode?

**Comment by de-nordic** [comment]
@Laczen I have incorporated your suggestion regarding adding information on image destination in header have delivered prototype solution that allows to sign image for mcumgr, with imgtool, with desired address set. This way image may be rejected if it is supposed to be uploaded at specified ROM address but selected slot will have different address.

**Comment by de-nordic** [comment]
> 
> 
> > 1. How MCUboot will handle 3 or 4 images, or even more?
> 
> The `mcumgr` CLI app already has a flag to choose the image pair where to upload, `image upload -n <image-number>`, which can be used to choose between S/NS, but this currently uploads in a fixed slot. I don't remember if the mcumgr library supports this, but MCUboot surely does. With the XIP support it will probably also get an extra flag for the slot number. Then a user will have control of what image pair and slot a particular image should be uploaded to.

The zephyr backend, of mcumgr, will attempt to select one of two images, which basically would be the non-active one.

**Comment by tejlmand** [comment]
@de-nordic please take a look here: #40555 I believe the approach in that RFC will be very suitable for this kind of work you need to do here.

**Comment by de-nordic** [comment]
> @de-nordic please take a look here: #40555 I believe the approach in that RFC will be very suitable for this kind of work you need to do here.

Thanks, I will take a look.

**Pull request closed without merging**